### ISSUE TITLE
Burnout Paradise, crash at start after intro video

### ISSUE THREAD

**Issue by kprkpr** [Joshua-Ashton/d9vk#167]
Hi
After starting Burnout Paradise, starts video, loads first menu, and sometimes starts loading samegame and crashes
I saw that if I wait, crashes in menu, some "deferred" thing?

### Software information
Burnout Paradise, stock settings (1366x768 and low-medium settings, I can't change it)
It works without D9VK but is slow

### System information
- GPU: Intel HD 4000 (I5 3320M)
- Driver: Mesa 19.0.6
- Wine version: wine-staging 4.9
- D9VK version: 0.12

### Log files
-[winefail.log](https://github.com/Joshua-Ashton/d9vk/files/3283310/winefail.log)
- [BurnoutParadise_d3d9.log](https://github.com/Joshua-Ashton/d9vk/files/3283282/BurnoutParadise_d3d9.log)

[BurnoutParadise_d3d9_debug.log](https://github.com/Joshua-Ashton/d9vk/files/3283307/BurnoutParadise_d3d9_debug.log)

**Comment by Kron4ek** [comment]
And it also crashes on next runs because of the game's shader cache (wineprefix/drive_c/users/USER/Local Settings/Application Data/Criterion Games/Burnout Paradise/ShaderCache). Removing ShaderCache directory allowing the game to launch again.

**Comment by kprkpr** [comment]
I ran again deleting shadercache (wineprefix/drive_c/users/USER/Local Settings/Application Data/Criterion Games/Burnout Paradise/ShaderCache) like _Kron? (I dont see comment now)_ saw me, and grab a debug again. It crashes in the same part but maybe is better log
[BurnoutParadise_d3d9.log](https://github.com/Joshua-Ashton/d9vk/files/3288012/BurnoutParadise_d3d9.log)

**Comment by kprkpr** [comment]
I tried to do with D9VK master and Mesa-git (Shows 19.1.99) and game starts and runs, but there is very hard glitches
![DeepinScreenshot_xfdesktop_20190613234050](https://user-images.githubusercontent.com/5132053/59469357-e0a71280-8e34-11e9-8365-1b4ca22133f1.png)
[BurnoutParadise_d3d9.log](https://github.com/Joshua-Ashton/d9vk/files/3288193/BurnoutParadise_d3d9.log)

This file is wine terminal when I closed the game (It crashed when I pressed "Save and Exit")
[somelog.txt](https://github.com/Joshua-Ashton/d9vk/files/3288196/somelog.txt)

**Comment by K0bin** [comment]
It fails to create a bunch of textures and chokes on a couple of shader instructions.

Please make an apitrace with WineD3D (or even better: Windows) and upload that.

**Comment by kprkpr** [comment]
Hi
I attach d3d9 .trace file with Windows 10 in same machine. I used ""D:\XX\apitrace-msvc-latest\apitrace-msvc\x86\bin\apitrace.exe" trace --api d3d9 BurnoutParadise.exe"

Is very big (1.2Gb) and I compressed it in 7z (+-250Mb) and uploaded to Mediafire http://www.mediafire.com/file/spde9nrhz97i2u6/BurnoutParadiseapitrace.7z

**Comment by K0bin** [comment]
Nice, but next time you upload a trace, pls upload it to a hoster that doesn't limit download to 250kb/s. Something like Google Drive, Firefox Send or Mega typically works well.

**Comment by kprkpr** [comment]
I reuploaded to Google Drive, sorry

Then.. Is driver problem? :|

**Comment by K0bin** [comment]
It's a hardware problem. Old Intel GPUs dont fully support Vulkan. Nothing that can be done there.

**Comment by kprkpr** [comment]
Hi. I uploaded bug report to mesa and now supports D24S8 in mesa-git (https://bugs.freedesktop.org/show_bug.cgi?id=99493 -> https://gitlab.freedesktop.org/mesa/mesa/merge_requests/1109). I updated it to master. apitrace of Windows seems to retrace right in wine qapiretrace with d9vk, but if I start game, although it renders better, doesn't work right. Works a bit better if I use ShaderCache folder using before WineD3D but not ideal. I don't know how to trace in wineD3D or something like that

New d3d9 log and attach images
![DeepinScreenshot_xfdesktop_20190618185412](https://user-images.githubusercontent.com/5132053/59703751-e026c700-91fa-11e9-8af7-887fefbff8b7.png) (Tires doesn't show)
![DeepinScreenshot_xfdesktop_20190618185428](https://user-images.githubusercontent.com/5132053/59703756-e452e480-91fa-11e9-92ed-686a87f2bf1c.png) (Buildings show "in bones")


[BurnoutParadise_d3d9.log](https://github.com/Joshua-Ashton/d9vk/files/3302593/BurnoutParadise_d3d9.log)

And some "loading screens" shows in black, in Windows/WineD3D shows an splash image of game

**Comment by K0bin** [comment]
```
err:   DxsoDecodeContext::decodeSourceRegister: source register out of range.
err:   DxsoDecodeContext::decodeSourceRegister: source register out of range.
err:   DxsoDecodeContext::decodeSourceRegister: source register out of range.
err:   DxsoDecodeContext::decodeSourceRegister: source register out of range.
```

Looks like the D9VK shader compiler has issues with some of the shaders. This is probably something we can fix on our end. I don't know if that's what causes your issues because the apitrace replays fine on Nvidia hardware with D9VK.

**Comment by kprkpr** [comment]
Windows Apitrace renders ok in d9vk in intel, but game doesnt runs as apitrace..
![DeepinScreenshot_Seleccionar área_20190618191940](https://user-images.githubusercontent.com/5132053/59705178-2893b400-91fe-11e9-8cd4-13d3a552ba7f.png)
![DeepinScreenshot_Seleccionar área_20190618191957](https://user-images.githubusercontent.com/5132053/59705179-2893b400-91fe-11e9-8ea6-772dd5320a4e.png)

**Comment by K0bin** [comment]
So your saying that a apitrace recorded on Windows replays fine on your hardware with D9VK but the game itself doesnt?

That sounds a bit suspicious. I own the game, gonna take a look.

**Comment by kprkpr** [comment]
Yes, it's like that.
Only difference is that savegame is not same, but it's recorded in same computer (Dual boot) and using WineD3D works (Its slow but work)
Using mesa-git of today with z24S8 working:
-Windows apitrace in D9VK: Render ok
-Game in WineD3D: Render ok
-Game in D9VK: Screenshots above..
Its strange.. Some card detection or like that? I'm not advanced in wine at so..

You have the game? Great ;)

**Comment by K0bin** [comment]
I can reproduce your issues with D9VK. The reason your trace is fine is probably because the game was running with higher graphics settings.
It seems to work at medium and higher. Turning everything to low breaks it.

**Comment by kprkpr** [comment]
I tried to turn on medium quality in Wine and same problem, no tires and half of buildings.. 
Later, I make an apitrace in Windows at low settings and replay with qapitrace in linux and replay well (Only black screen in loading vs "Burnout paradise loading screen")

**Comment by K0bin** [comment]
High settings work fine with D9VK,
Can you post your apitrace? I can't even get it to run on Windows with low settings. It just crashes.

**Comment by kprkpr** [comment]
For testing, I tried with my workstation (Intel I5 4590, AMD RX580, d9vk 0.12) and...
1st time loading game -> Was in Medium settings, all seemed right. Change to max settings, crash game
2nd time loading game (After this crash) -> Same as intel gpu, no tires and problems with buildings...

Maybe is another thing, maybe some cache of d9vk/vulkan?

**Comment by K0bin** [comment]
> Maybe is another thing, maybe some cache of d9vk/vulkan?

No.
As I said, it crashes with Windows D3D9 on low settings for me.

**Comment by kprkpr** [comment]
Low apitrace (With intel): https://drive.google.com/open?id=1JwfFAhH0JBfXqyix3VcJogfTJqXplM6e

Sometimes, I can't start game until I delete "wineprefix/drive_c/users/USER/Local Settings/Application Data/Criterion Games/Burnout Paradise/ShaderCache" folder (Instant crash at startup). I don't know if its related with that, attached part of winelog (Its same in AMD and Intel..)
[wineloginstantcrash.log](https://github.com/Joshua-Ashton/d9vk/files/3303670/wineloginstantcrash.log)

**Comment by K0bin** [comment]
Are you using Microsofts d3dx9_37 dll? The Wine one doesnt really work.

**Comment by K0bin** [comment]
![grafik](https://user-images.githubusercontent.com/1131720/59725872-a328f780-922f-11e9-9dd2-6d4109f5f388.png)

That trace also works fine with D9VK.

**Comment by kprkpr** [comment]
After you say that, I enabled in winecfg d3dx9_37 as "Native (Windows)", but same..
![8cf1a294-9148-4c12-831a-a732a23abc9c](https://user-images.githubusercontent.com/5132053/59742314-b490f480-926d-11e9-8cdc-2ca521c83f10.jpg)

**Comment by Kron4ek** [comment]
I have the same issue with invisible wheels and buildings and some other objects. And the game sometimes crashes until ShaderCache directory removed. Seems like the game creates wrong shaders cache with D9VK. The game renders correctly if ShaderCache was created using WineD3D.

How to make the game renders correctly:

1, Remove ShaderCache directory
2. Run the game using WineD3D and let it fullfill ShaderCache directory
3. After that run the game using D9VK
4. It renders correctly now, because ShaderCache was correctly created using WineD3D

How to make the game renders incorrectly:

1. Remove ShaderCache directory
2. Run the game using D9VK
3. Now it creates wrong ShaderCache and renders incorrectly

There are many dxso errors when ShaderCache created using D9VK.

    err:   DxsoDecodeContext::decodeSourceRegister: source register out of range.

But there are no such errors when ShaderCache created using WineD3D.

I have and Intel HD 630 and Mesa 19.1.0, and i use d3dx9 and d3dcompiler libraries from Windows.

**Comment by K0bin** [comment]
It makes absolutely zero sense that cached WineD3D shaders would impact D9VK at all. It's a lot more likely that it just works out of dumb luck.

**Comment by Kron4ek** [comment]
Well, i tested many times (10+ runs), and it always renders incorrectly if ShaderCache was created with D9VK. And always renders correctly 

And graphics settings doesn't make any effect for me, i tried low, medium and high presets and the result is the same for all of them.

> cached WineD3D shaders

I mean game's shader cache which is in Local Settings/Application Data/Criterion Games/Burnout Paradise/ShaderCache.

**Comment by K0bin** [comment]
> I mean game's shader cache which is in Local Settings/Application Data/Criterion Games/Burnout Paradise/ShaderCache.

Ok, I thought you meant the drivers shader cache. No idea what that does. Interesting.

**Comment by Kron4ek** [comment]
Here is the video: https://youtu.be/G4zyykh5WOU

**Comment by K0bin** [comment]
So the game actually stores D3D9 shader byte code inside that shader cache folder. It retrieves that bytecode using D3D9*Shader->GetFunction(). We've found an issue with this function in D9VK.

**Comment by K0bin** [comment]
Fixed in 0e6e448d0f9a5f2abc6245a99575460a624b54b6

Keep in mind that this commit currently only exists in the `d3d9-fixed-function` branch.

**Comment by Kron4ek** [comment]
It works good now, thanks @Joshua-Ashton and @K0bin.

**Comment by kprkpr** [comment]
Tested and perfect! 45-60fps at low (WineD3D was..20fps) and solved render issues and now doesn't crash when exit too
Thanks @Joshua-Ashton @K0bin and @Kron4ek
